File: Haskell/hoshimeguri1.hs
--  ビルド:
--    $ ghc --make hoshimeguri1.hs
--
--  実行:
--    $ ./hoshimeguri1 ../data.txt

import System.Environment
import System.IO

star 'A' 'W' = 'I'
star 'A' 'R' = 'H'
star 'I' 'W' = 'G'
star 'I' 'R' = 'F'
star 'G' 'W' = 'E'
star 'G' 'R' = 'D'
star 'E' 'W' = 'C'
star 'E' 'R' = 'B'
star 'C' 'W' = 'A'
star 'C' 'R' = 'J'
star 'H' 'W' = 'C'
star 'H' 'R' = 'J'
star 'J' 'W' = 'E'
star 'J' 'R' = 'B'
star 'B' 'W' = 'G'
star 'B' 'R' = 'D'
star 'D' 'W' = 'I'
star 'D' 'R' = 'F'
star 'F' 'W' = 'A'
star 'F' 'R' = 'H'

solve :: String -> String
solve input =
  reverse $ foldl (\a c -> (star (head a) c):a) [head input] (tail input)

test :: [String] -> IO ()
test [input, expected] =
  if actual == expected
    then putStr "."
    else putStrLn $ concat
      [ "\ninput:    ", input
      , "\nexpected: ", expected
      , "\nacutal:   ", actual
      ]
  where
    actual = solve input

main = getArgs >>= return.head >>= (flip $ openFile) ReadMode >>= hGetContents >>= mapM_ (test.words).lines >> putStrLn ""

File: Rust/hoshimeguri1.rs
//  ビルド:
//    $ rustc hoshimeguri1.rs
//
//  実行:
//    $ ./hoshimeguri1 ../data.txt

fn star(p: char, c: char) -> char {
    match (p, c) {
        ('A', 'W') => 'I',
        ('A', 'R') => 'H',
        ('I', 'W') => 'G',
        ('I', 'R') => 'F',
        ('G', 'W') => 'E',
        ('G', 'R') => 'D',
        ('E', 'W') => 'C',
        ('E', 'R') => 'B',
        ('C', 'W') => 'A',
        ('C', 'R') => 'J',
        ('H', 'W') => 'C',
        ('H', 'R') => 'J',
        ('J', 'W') => 'E',
        ('J', 'R') => 'B',
        ('B', 'W') => 'G',
        ('B', 'R') => 'D',
        ('D', 'W') => 'I',
        ('D', 'R') => 'F',
        ('F', 'W') => 'A',
        ('F', 'R') => 'H',
        _          => '_'
    }
}

fn solve(input: &str) -> String {
    let mut colors = input.chars();
    let mut p = colors.next().unwrap();
    let mut output = String::new();
    output.push(p);
    for c in colors {
        p = star(p, c);
        output.push(p)
    }
    output
}

fn test(input: &str, expected: &str) {
    let actual = solve(input);
    if actual == expected {
        print!(".")
    } else {
      println!(r#"
input:    {}
expected: {}
actual:   {}"#, input, expected, actual)
    }
}

fn read_lines(filename: String, lines: &mut Vec<String>) -> std::io::Result<()> {
    use std::io::prelude::*;
    use std::fs::File;
    let mut file = try!(File::open(filename));
    let mut contents = String::new();
    try!(file.read_to_string(&mut contents));
    for line in contents.lines() {
        lines.push(line.to_string())
    }
    Ok(())
}

fn main() {
    use std::env;

    let mut args = env::args();

    match args.nth(1) {
        Some(filename) => {
            let mut lines: Vec<String> = Vec::new();
            read_lines(filename, &mut lines);
            for line in lines {
                let mut words = line.split_whitespace();
                let input = words.next();
                let expected = words.next();
                test(input.unwrap(), expected.unwrap())
            }
        }
        None => println!("no file")
    }
    println!("")
}
